tests/dragonfly/conftest.py

```diff
@@ -254,30 +254,22 @@ def memcached_connection(df_server: DflyInstance):
 
 
 @pytest.fixture(scope="session")
-def gen_ca_cert(tmp_dir):
-    # We first need to generate the tls certificates to be used by the server
-
-    # Generate CA (certificate authority) key and self-signed certificate
-    # In production, CA should be generated by a third party authority
-    # Expires in one day and is not encrtypted (-nodes)
-    # X.509 format for the key
+def with_tls_ca_cert_args(tmp_dir):
     ca_key = os.path.join(tmp_dir, "ca-key.pem")
     ca_cert = os.path.join(tmp_dir, "ca-cert.pem")
-    step = rf'openssl req -x509 -newkey rsa:4096 -days 1 -nodes -keyout {ca_key} -out {ca_cert} -subj "/C=GR/ST=SKG/L=Thessaloniki/O=KK/OU=AcmeStudios/CN=Gr/emailAddress=acme@gmail.com"'
-    subprocess.run(step, shell=True)
-
+    gen_ca_cert(ca_key, ca_cert)
```

Comment:
This was to make sure the client fails to connect with the old CA - if we reused the CA the client could still connect even if the TLS certs weren't properly updated?

Reply:
> if we reused the CA the client could still connect even if the TLS certs weren't properly updated

The `certified authority -- ca` is used to verify the authenticity of the client. How does the browser work then on `https` request? Do you `sign` your certificates via a `CA` which you then send to the end server(like google.com) to load it for you such that it can auth ? 

What I am trying to say here, `CA` is dose not control who connects to the server, `CA` is just a safety step that the certificate advertised by a client is legit. TLS just encrypts the trafic such that evedrops etc don't happen. It's not a security measure against who has the right to connect to a service. For that you would need a login/password setup. 

> if we reused the CA the client could still connect even if the TLS certs weren't properly updated so I'm not sure how we'd verify the server certs updated?

If you used a global `CA` or the one supplied by the OS you would be surprised that in production you would be able to allow more than you bargain for -- and that's intended -- because a) that's how the WWW works b) TLS is just used for encryption between connections and not as a mechanism to auth users :)

Reply:
One more thing, if I still the exchanged between a DF instance and the client, they will be useless, because I would not have the private key required to decrypt the message

Reply:
I hope this is helpful, I am not a security expert but this should be the gist of it

Reply:
> is used to verify the authenticity of the client

It is used to verify any certificate - both client and server.

In test `test_config_update_tls_certs` the CA cert is only being used by the client to verify the server certificate (if we didn't specify the CA cert it would use system defaults, but given the server is using a self signed cert the client must be configured with the CA or it will reject the servers cert). The client doesn't have certificates.

Such as with Redis CLI:
```
$ redis-cli --tls ping
Could not connect to Redis at 127.0.0.1:6379: SSL_connect failed: certificate verify failed
$ redis-cli --tls --cacert ca-cert.pem ping
PONG
$ redis-cli --tls --cacert ca-cert-old.pem ping
Could not connect to Redis at 127.0.0.1:6379: SSL_connect failed: certificate verify failed
```

So my point is the `test_config_update_tls_certs` must verify the server is using the new certificates. Though if the new certificates were signed by the same CA, the client would connect successfully even if Dragonfly was actually still using the old certificates?

Therefore the test generates a new CA for the new certificates then:
* A client using the old CA MUST fail to verify the server given the server certificates should be signed by the new CA
* A client using the new CA MUST connect

I agree generating another CA is a bit of a pain, though I'm not sure how else to verify that Dragonfly is using the new certificates?

Reply:
> Though if the new certificates were signed by the same CA, the client would connect successfully even if Dragonfly was actually still using the old certificates?

Assuming, you are using an authorized CA, then any new certificate (public key) can be authorized by the CA root. Btw, when you are saying `certificates` what I understand is `public key and private key`. The CA  itself is different.

> I agree generating another CA is a bit of a pain, though I'm not sure how else to verify that Dragonfly is using the new certificates?

You generate a new client public/private key which the server auths again the new CA root. Since you used a new CA root, it can only authorize certificates(public keys) that it signed. Therefore, when you reload the CA, DF should reject the old public keys since they were not signed by it (the CA is different from the previous one and hence, old clients certificates won't be validated on the server side/DF)

Reply:
oh oh and that's why you wanted the new CA. NVM -- you already do that

Reply:
Honestly, it doesn't have to be bidirectional. What you do is more than enough. It's really not worth it to add what I just described since what you did covers the correctness side of things -- so basically what I am saying is noise :)